File: frontend/src/api/analysis.ts
// src/api/analysis.ts

export interface EnvironmentBreakdown {
  name: string;
  count: number;
}

export interface OSSummary {
  family: string;
  count: number;
  legacy_count: number;
}

export interface ReadinessSummary {
  ready_for_rehost: number;
  needs_modernization: number;
  needs_investigation: number;
}

export interface AnalysisSummary {
  run_id: string;
  total_servers: number;
  environments: EnvironmentBreakdown[];
  os_summary: OSSummary[];
  legacy_server_count: number;
  modern_server_count: number;
  readiness: ReadinessSummary;
  notes: string[];
}

export interface RecommendationItem {
  server_id: string;
  hostname: string | null;
  ip: string | null;
  environment: string | null;
  os: string | null;
  strategy: string;
  risk_level: string;
  wave: number;
  summary: string;
  estimated_monthly_savings: number | null;
  target_platform: string | null;
}

export interface AnalysisRecommendations {
  run_id: string;
  total_recommendations: number;
  items: RecommendationItem[];
}

// IMPORTANT: match how nginx is already proxying the backend.
// Existing working calls (run registry, etc.) go to /v1/..., not /api/v1/...
const API_BASE = "/v1/analysis";

async function handleResponse<T>(res: Response): Promise<T> {
  if (!res.ok) {
    let detail: string | undefined;
    try {
      const data = await res.json();
      detail = (data as any)?.detail;
    } catch {
      // ignore JSON parse errors; fall back to status text
    }
    throw new Error(
      detail || `Request failed with status ${res.status} ${res.statusText}`
    );
  }
  return (await res.json()) as T;
}

export async function fetchAnalysisSummary(
  runId: string
): Promise<AnalysisSummary> {
  const res = await fetch(
    `${API_BASE}/${encodeURIComponent(runId)}/summary`
  );
  return handleResponse<AnalysisSummary>(res);
}

export async function fetchAnalysisRecommendations(
  runId: string
): Promise<AnalysisRecommendations> {
  const res = await fetch(
    `${API_BASE}/${encodeURIComponent(runId)}/recommendations`
  );
  return handleResponse<AnalysisRecommendations>(res);
}

File: backend/app/schemas/analysis.py
from typing import List, Optional
from pydantic import BaseModel


class EnvironmentBreakdown(BaseModel):
    """
    Simple count of servers per environment (e.g., Prod, Dev, Test).
    """
    name: str
    count: int


class OSSummary(BaseModel):
    """
    Aggregated view of OS families and how many are legacy/EOL.
    """
    family: str
    count: int
    legacy_count: int


class ReadinessSummary(BaseModel):
    """
    High-level readiness banding for the run.
    """
    ready_for_rehost: int
    needs_modernization: int
    needs_investigation: int


class RecommendationItem(BaseModel):
    """
    Per-server recommendation row.
    """
    server_id: str
    hostname: Optional[str]
    ip: Optional[str]
    environment: Optional[str]
    os: Optional[str]

    strategy: str              # e.g., "Rehost", "Refactor / Modernize"
    risk_level: str            # e.g., "Low", "Medium", "High"
    wave: int                  # migration wave number

    summary: str               # short explanation

    estimated_monthly_savings: Optional[float] = None
    target_platform: Optional[str] = None


class AnalysisSummary(BaseModel):
    """
    Top-level summary returned by /v1/analysis/{run_id}/summary
    """
    run_id: str

    total_servers: int
    environments: List[EnvironmentBreakdown]
    os_summary: List[OSSummary]

    legacy_server_count: int
    modern_server_count: int

    readiness: ReadinessSummary

    notes: List[str]


class AnalysisRecommendations(BaseModel):
    """
    Canonical response model for /v1/analysis/{run_id}/recommendations
    """
    run_id: str
    total_recommendations: int
    items: List[RecommendationItem]


# --- Backward-compatible aliases -------------------------------------------

class RecommendationResponse(AnalysisRecommendations):
    """
    Alias used by older code paths (run_views_v2 / run_views_servers_v2).
    """
    pass


class RecommendationsResponse(AnalysisRecommendations):
    """
    Alias used by routers expecting RecommendationsResponse.
    """
    pass
